### ISSUE TITLE
Minor bug?

### ISSUE THREAD

**Issue by gnzlbg** [Manu343726/cpp-dod-tests#1]
Should

https://github.com/Manu343726/cpp-dod-tests/blob/master/dod_perfv1.cpp#L87

say `a.p - p` instead of `a - p - p` ?

**Comment by Manu343726** [comment]
Let's make @joaquintides enter the discussion, since he's the author of those tests.

**Comment by Manu343726** [comment]
But you are right, it seems to be a little typo. Anyway, the latest version (`dod_perfv2`) may have that fixed. v1 was included in the matrix to compare changes and fixes.

**Comment by joaquintides** [comment]
You're right, this is a typo I didn't bother to fix because the code is anyway superseded by v2, which is much faster. 

Are you testing the thing? Any interesting results?

**Comment by gnzlbg** [comment]
I started working on testing it today, I will probably have some first
results over the weekend.

I gave it a lot of thought today and the main doubt I still have if it is
worth it at all to make this generic. At the end of the day it is just a
struct with pointers, and depending on the data layout, some pointers might
need an offset: e.g. suppose the dx and dy are in different arrays like in
your example, but the x and y are in the same array interleaved like x, y,
x, y, ... Then there is "no need" to use a pointer for x and a pointer for
y, one can use the same pointer for both, but an offset is required.
Another thing is that writing `using...` + `get(x())` is a pain, so one
would probably wrap the `get(...)` into its own functions to save some
typing. At that point rolling the struct by hand or using the generic
approach is more a matter of taste I guess.

Still I've been only thinking about this for half a day now. I will push it
and see what happens.

**Comment by joaquintides** [comment]
I'd say the main benefit of abstracting access away is that you can easily change layout at will without touching the derived class at all. For instance, it's very easy to augment the access implementation provided to support interleaved (in the sense you describe) members: just partially specialize for access<interleaved<Members...>,MoreMembers...> and write the machinery to support multiple get() overloads based in just one base address. (If you find difficulties figuring this out just let me know and I can sketch something in more detail.) Another, probably less important, benefit is that you can have "pointers" granting access only to members needed by foo and, if you need to add new members accessed by some bar member function then the code using foo need not change.

As for the clunky get(x()) syntax, yes, I'm not very comfortable with that myself either, in particular about the () thing after x. If you find something terser I'd love to hear about it.

**Comment by Manu343726** [comment]
> As for the clunky get(x()) syntax, yes, I'm not very comfortable with that myself either, in particular about the () thing after x. If you find something terser I'd love to hear about it.

What about `constexpr` variables?

``` cpp
template<typename Access>
class particle : Access
{
    static constexpr auto x = member<int,0>{};
    static constexpr auto dx = member<int,2>{};
    void move()
    {
        get(x)+=get(dx);
    }
};

```

Also you can wrap the `get()`er with an assignment proxy, giving an interface similar to a struct, `particle.x = 0` etc. See https://github.com/Manu343726/octopus/blob/master/blocks/manu343726/math/vector.hpp#L533

**Comment by gnzlbg** [comment]
I think I expressed myself wrong. I wasn't arguing about how important it is to abstract access away (I agree it is very important, it is actually the whole point of this). 

To show what I meant just comparing the following particle with the one using Access [here](https://github.com/Manu343726/cpp-dod-tests/blob/master/dod_perfv2.cpp#L214): 

``` c++
struct particle {

private:
  char* color_;
  int*  x_;
  int*  y_;
  int*  dx_;
  int*  dy_;

public:
  char& color() { return *color_; }
  char const& color() const { return *color_; }
  int& x() { return *x_; }
  int const& x() const { return *x_; }
  int& y() { return *y_; }
  int const& y() const { return *y_; }

  int& dx() { return *dx_; }
  int const& dx() const { return *dx_; }
  int& dy() { return *dy_; }
  int const& dy() const { return *dy_; }

  void render() {
    do_render(x(), y(), color());
  }

  void move() {
    x()+=dx();
    if(x()<0){
      x()*=-1;
      dx()*=-1;
    }
    else if(x()>max_x){
      x()=2*max_x-x();
      dx()*=-1;      
    }

    y()+=dy();
    if(y()<0){
      y()*=-1;
      dy()*=-1;
    }
    else if(y()>max_y){
      y()=2*max_y-y();
      dy()*=-1;      
    }
  }

  particle(...);
};
```

Code that works on particle just uses the member functions. How particle is actually implemented or where the data lives is "abstracted away" from the user of particle. We could easily change particle to use an [x,y,x,...] layout by just doing this:

``` c++
struct particle {

private:
  int*  xs_;
  /// ...

public:

  int& x() { return xs_[0]; }
  int const& x() const { return xs_[0]; }
  int& y() { return xs_[1]; }
  int const& y() const { return xs_[1]; }
  /// ...

};
```

User code doesn't need to change at all (this is what I mean when I said this abstracts access). 

We could also use simple struct inheritance to reuse this particle. 

So basically, given that this is possible in current C++, easy to do, and "pretty clear", my single point was that I am not sure about "what do the `Access`/`Member`/`tags`/`make_pointer`, ... buy us over this approach, and if it's worth it.

**Comment by joaquintides** [comment]
Yes, constexpr's work nicely here --not so much when typedefing access<...>, where the args must be types and so x, dx etc can't be used. 

Not so sure about proxies, they interfere with generic code.

**Comment by joaquintides** [comment]
Gnzlbg: expanding on your approach, how would you implement particle traversal? Note that the access framework I provide gives you that for free without extra coding.

**Comment by gnzlbg** [comment]
Right, I just make particle model a random access iterator, in range-v3
that is actually pretty easy but one has to do so manually for each "type".

In your approach you get it for free unless you need something more special
(strides, tiling with halo particles, ...)., in which case you need special
access functions, each with their own stride probably. What if you want to
reuse a stride for two pointers? Unless one makes the stride a template
parameter one needs to store it twice :/

**Comment by joaquintides** [comment]
Your particle does not model a random access iterator: you need either make it do so or write an auxiliary particle_iterator class. Maybe I'm not getting your point, I suggest you provide some (pseudo)code on how particle traversal would look like in your view. 

As for strides, I'm almost sure I'm not getting your point either :-) but seems to me in my approach you just need advance pointer<particle>'s by n, right?
